smrf/envphys/solar/toporad.py

```diff
@@ -0,0 +1,285 @@
+from topocalc.horizon import horizon
+from topocalc.shade import shade
+
+from smrf.envphys.solar.twostream import twostream
+from smrf.envphys.solar.irradiance import direct_solar_irradiance
+from smrf.envphys.thermal.topotherm import hysat
+from smrf.envphys.albedo import albedo
+from smrf.envphys.constants import SEA_LEVEL, STD_LAPSE, \
+    GRAVITY, MOL_AIR, STD_AIRTMP
+
+VISIBLE_MIN = .28
+VISIBLE_MAX = .7
+IR_MIN = .7
+IR_MAX = 2.8
+
+
+def check_wavelengths(wavelength_range):
+
+    if wavelength_range[0] >= VISIBLE_MIN and \
+            wavelength_range[1] <= VISIBLE_MAX:
+        wavelength_flag = 'vis'
+    elif wavelength_range[0] >= IR_MIN and wavelength_range[1] <= IR_MAX:
+        wavelength_flag = 'ir'
+    else:
+        raise ValueError(
+            'stoporad wavelength range not within visible or IR wavelengths')
+
+    return wavelength_flag
+
+
+def stoporad(date_time, topo, cosz, azimuth, illum_ang, albedo_surface,
+             wavelength_range, tau_elevation=100, tau=0.2, omega=0.85,
+             scattering_factor=0.3):
+
+    wavelength_flag = check_wavelengths(wavelength_range)  # noqa
+
+    # check cosz if sun is down
+    if cosz < 0:
+        return np.zeros_like(topo.dem), np.zeros_like(topo.dem)
+
+    else:
+        solar_irradiance = direct_solar_irradiance(
+            date_time, w=wavelength_range)
+
+        # Run horizon to get sun-below-horizon mask
+        horizon_angles = horizon(azimuth, topo.dem, topo.dx)
+        thresh = np.tan(np.pi / 2 - np.arccos(cosz))
+        no_sun_mask = np.tan(np.abs(horizon_angles)) > thresh
+
+        # Run shade to get cosine local illumination angle
+        # mask by horizon mask using cosz=0 where the sun is not visible
+        illum_ang = np.copy(illum_ang)
+        illum_ang[no_sun_mask] = 0
+
+        R0 = np.mean(albedo_surface)
+
+        # Run elevrad to get beam & diffuse then toporad
+        evrad = Elevrad(
+            topo.dem,
+            solar_irradiance,
+            cosz,
+            tau_elevation=tau_elevation,
+            tau=tau,
+            omega=omega,
+            scattering_factor=scattering_factor,
+            surface_albedo=R0)
+
+        trad_beam, trad_diff = toporad(
+            evrad.beam,
+            evrad.diffuse,
+            illum_ang,
+            topo.sky_view_factor,
+            topo.terrain_config_factor,
+            cosz,
+            surface_albedo=albedo_surface)
+
+    return trad_beam, trad_diff
+
+
+def stoporad_ipw(date_time, tau_elevation, tau, omega, scattering_factor,
+                 wavelength_range, start, current_day, time_zone,
+                 year, cosz, azimuth, grain_size, max_grain, dirt, topo):
+    """stoporad simulates topographic radiation over snow-covered terrain.
+    Uses a two-stream atmospheric radiation model. This function mimics the
+    original IPW stoporad program and differences are bit noise resolution
+    of the 8-bit IPW images.
+
+    TODO: depricated, mainly for testing
```

Comment:
Typo: deprecated

(Spelling found in more places)

Reply:
Dang spell checker didn't catch this...